internal/worker/apiaddresssetter/worker.go

```diff
@@ -0,0 +1,367 @@
+func (w *apiAddressSetterWorker) Wait() error {
+	return w.catacomb.Wait()
+}
+
+func (w *apiAddressSetterWorker) loop() error {
+	controllerNodeChanges, err := w.watchForControllerNodeChanges()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	configChanges, err := w.watchForConfigChanges()
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	ctx, cancel := w.scopedContext()
+	defer cancel()
+
+	for {
+		w.config.Logger.Tracef(ctx, "waiting for controller nodes or addresses changes")
+		select {
+		case <-w.catacomb.Dying():
+			return w.catacomb.ErrDying()
+
+		case <-controllerNodeChanges:
+			// A controller was added or removed.
+			w.config.Logger.Tracef(ctx, "<-controllerNodeChanges")
+			changed, err := w.updateControllerNodes(ctx)
+			if err != nil {
+				return errors.Capture(err)
+			}
+			if !changed {
+				continue
+			}
+			w.config.Logger.Tracef(ctx, "controller node added or removed")
+
+		case <-w.controllerNodeChanges:
+			// One of the controller nodes changed.
+			w.config.Logger.Tracef(ctx, "<-w.controllerNodeChanges")
+
+		case <-configChanges:
+			// Controller config has changed.
+			w.config.Logger.Tracef(ctx, "<-w.configChanges")
+
+			if len(w.runner.WorkerNames()) == 0 {
+				w.config.Logger.Errorf(ctx, "no controller information, ignoring config change")
+				continue
+			}
+		}
+
+		if err := w.updateAPIAddresses(ctx); err != nil {
+			w.config.Logger.Errorf(ctx, "cannot update api addresses: %v", err)
+		}
+	}
+}
+
+func (w *apiAddressSetterWorker) scopedContext() (context.Context, context.CancelFunc) {
+	return context.WithCancel(w.catacomb.Context(context.Background()))
+}
+
+// watchForControllerChanges starts a watcher for changes to controller nodes.
+// It returns a channel which will receive events if any of the watchers fires.
+func (w *apiAddressSetterWorker) watchForControllerNodeChanges() (<-chan struct{}, error) {
+	watcher, err := w.config.ControllerNodeService.WatchControllerNodes()
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	if err := w.catacomb.Add(watcher); err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	return watcher.Changes(), nil
+}
+
+// watchForConfigChanges starts a watcher for changes to controller config.
+// It returns a channel which will receive events if the watcher fires.
+func (w *apiAddressSetterWorker) watchForConfigChanges() (<-chan []string, error) {
+	watcher, err := w.config.ControllerConfigService.WatchControllerConfig()
+	if err != nil {
+		return nil, errors.Capture(err)
+	}
+	if err := w.catacomb.Add(watcher); err != nil {
+		return nil, errors.Capture(err)
+	}
+
+	return watcher.Changes(), nil
+}
+
+// updateControllerNodes updates the current list of tracked controller nodes,
+// as well as starting and stopping trackers for them as they are added and
+// removed.
+func (w *apiAddressSetterWorker) updateControllerNodes(ctx context.Context) (bool, error) {
+	controllerIDs, err := w.config.ControllerNodeService.GetControllerIDs(ctx)
```

Comment:
As an aside, if the get controller ids returns an error for zero controller ids, then this will bounce hard and fast. What depends on this?